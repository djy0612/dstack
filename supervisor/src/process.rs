use anyhow::{bail, Result};
use bon::Builder;
use fs_err as fs;
use notify::{RecursiveMode, Watcher};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::io::Write;
use std::marker::Unpin;
use std::path::Path;
use std::process::ExitStatus;
use std::process::Stdio;
use std::sync::MutexGuard;
use std::sync::{Arc, Mutex};
use std::time::SystemTime;
use tokio::io::{AsyncRead, AsyncReadExt};
use tokio::process::{Child, Command};
use tokio::sync::mpsc;
use tokio::sync::oneshot;
use tracing::Instrument;
use tracing::{error, info};

#[derive(Debug, Clone, Builder, Serialize, Deserialize)]
pub struct ProcessConfig {
    pub id: String,
    #[serde(default)]
    pub name: String,
    pub command: String,
    #[serde(default)]
    pub args: Vec<String>,
    #[serde(default)]
    pub env: HashMap<String, String>,
    #[serde(default)]
    pub cwd: String,
    #[serde(default)]
    pub stdout: String,
    #[serde(default)]
    pub stderr: String,
    #[serde(default)]
    pub pidfile: String,
    #[serde(default)]
    pub cid: Option<u32>,
    #[serde(default)]
    pub note: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessInfo {
    pub config: ProcessConfig,
    pub state: ProcessState,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessState {
    pub status: ProcessStatus,
    pub started: bool,
    pub pid: Option<u32>,
    #[serde(with = "systime")]
    pub started_at: Option<SystemTime>,
    #[serde(with = "systime")]
    pub stopped_at: Option<SystemTime>,
}

#[derive(Debug)]
pub(crate) struct ProcessStateRT {
    status: ProcessStatus,
    started: bool,
    pid: Option<u32>,
    kill_tx: Option<oneshot::Sender<()>>,
    started_at: Option<SystemTime>,
    stopped_at: Option<SystemTime>,
}

impl ProcessStateRT {
    pub(crate) fn is_running(&self) -> bool {
        self.status.is_running()
    }

    pub(crate) fn is_started(&self) -> bool {
        self.started
    }
}

impl ProcessStateRT {
    pub fn display(&self) -> ProcessState {
        ProcessState {
            status: self.status.clone(),
            started: self.started,
            pid: self.pid,
            started_at: self.started_at,
            stopped_at: self.stopped_at,
        }
    }
}

mod systime {
    use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
    use std::time::{Duration, SystemTime, UNIX_EPOCH};

    pub(crate) fn serialize<S: Serializer>(
        time: &Option<SystemTime>,
        serializer: S,
    ) -> Result<S::Ok, S::Error> {
        time.map(|t| t.duration_since(UNIX_EPOCH).unwrap().as_secs())
            .serialize(serializer)
    }

    pub(crate) fn deserialize<'de, D: Deserializer<'de>>(
        deserializer: D,
    ) -> Result<Option<SystemTime>, D::Error> {
        let Some(secs) = Option::<u64>::deserialize(deserializer)? else {
            return Ok(None);
        };
        let time = UNIX_EPOCH
            .checked_add(Duration::from_secs(secs))
            .ok_or_else(|| D::Error::custom("invalid unix timestamp"))?;
        Ok(Some(time))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProcessStatus {
    Running,
    Stopped,
    Exited(i32),
    Error(String),
}

impl ProcessStatus {
    pub fn is_running(&self) -> bool {
        matches!(self, ProcessStatus::Running)
    }

    pub fn is_stopped(&self) -> bool {
        matches!(self, ProcessStatus::Stopped)
    }
}

#[derive(Clone)]
pub(crate) struct Process {
    config: Arc<ProcessConfig>,
    state: Arc<Mutex<ProcessStateRT>>,
}

impl Process {
    pub fn new(config: ProcessConfig) -> Self {
        Self {
            config: Arc::new(config),
            state: Arc::new(Mutex::new(ProcessStateRT {
                pid: None,
                kill_tx: None,
                status: ProcessStatus::Stopped,
                started: false,
                started_at: None,
                stopped_at: None,
            })),
        }
    }

    pub(crate) fn lock(&self) -> MutexGuard<ProcessStateRT> {
        self.state.lock().unwrap()
    }

    pub fn start(&self) -> Result<()> {
        if self.lock().is_running() {
            bail!("Process is already running");
        }

        // Create command and spawn process
        let mut command = Command::new(&self.config.command);
        command
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .args(&self.config.args)
            .envs(&self.config.env)
            .kill_on_drop(true);
        if !self.config.cwd.is_empty() {
            command.current_dir(&self.config.cwd);
        }
        if !self.config.stdout.is_empty() {
            command.stdout(Stdio::piped());
        } else {
            command.stdout(Stdio::null());
        }
        if !self.config.stderr.is_empty() {
            command.stderr(Stdio::piped());
        } else {
            command.stderr(Stdio::null());
        }

        let mut process = command.spawn()?;
        let pid = process.id();

        let (kill_tx, kill_rx) = oneshot::channel();

        // Update process state
        {
            let mut state = self.state.lock().unwrap();
            state.started_at = Some(SystemTime::now());
            state.status = ProcessStatus::Running;
            state.pid = pid;
            state.kill_tx = Some(kill_tx);
            state.started = true;
        }

        // Handle IO redirection
        {
            let pidfile_path = self.config.pidfile.clone();
            if !pidfile_path.is_empty() {
                if let Err(err) =
                    fs_err::write(&pidfile_path, format!("{}", process.id().unwrap_or(0)))
                {
                    error!("Failed to write pidfile: {err}");
                }
            }

            let stdout = process.stdout.take();
            let stderr = process.stderr.take();
            let stdout_path = self.config.stdout.clone();
            let stderr_path = self.config.stderr.clone();

            if let Some(stdout) = stdout {
                tokio::spawn(redirect(stdout, stdout_path));
            }
            if let Some(stderr) = stderr {
                tokio::spawn(redirect(stderr, stderr_path));
            }
        }

        // Task for waiting on process
        {
            let process_uuid = self.config.id.clone();
            let weak_state = Arc::downgrade(&self.state);

            let span = tracing::info_span!("process", id = process_uuid);
            tokio::spawn(
                async move {
                    info!("Started");
                    let (killed, result) = wait_on_process(process, kill_rx).await;
                    let state = weak_state.upgrade();
                    let next_status = match result {
                        Ok(status) => {
                            if killed {
                                info!("Stopped");
                            } else if status.success() {
                                info!("Exited");
                            } else {
                                error!("Exited: {status:?}");
                            }
                            if killed {
                                ProcessStatus::Stopped
                            } else {
                                ProcessStatus::Exited(exit_code(status))
                            }
                        }
                        Err(e) => {
                            error!("Failed to wait on process: {e:?}");
                            ProcessStatus::Error(e.to_string())
                        }
                    };
                    if let Some(state) = state {
                        let mut state = state.lock().unwrap();
                        state.status = next_status;
                        state.stopped_at = Some(SystemTime::now());
                    }
                }
                .instrument(span),
            );
        }

        Ok(())
    }

    pub fn stop(&self) -> Result<()> {
        let mut state = self.state.lock().unwrap();
        state.started = false;
        let is_running = state.status.is_running();
        let Some(stop_tx) = state.kill_tx.take() else {
            if is_running {
                bail!("Missing kill tx for process");
            }
            return Ok(());
        };
        match stop_tx.send(()) {
            Ok(()) => Ok(()),
            Err(()) => match is_running {
                true => bail!("Failed to send stop signal to process"),
                false => Ok(()),
            },
        }
    }

    pub fn info(&self) -> ProcessInfo {
        let state = self.state.lock().unwrap();
        ProcessInfo {
            config: (*self.config).clone(),
            state: state.display(),
        }
    }
}

#[cfg(unix)]
fn exit_code(status: ExitStatus) -> i32 {
    use std::os::unix::process::ExitStatusExt;
    status.into_raw()
}

#[cfg(not(unix))]
fn exit_code(status: ExitStatus) -> i32 {
    status.code().unwrap_or(0)
}

async fn wait_on_process(
    mut process: Child,
    kill_rx: oneshot::Receiver<()>,
) -> (bool, Result<ExitStatus>) {
    let (killed, result) = tokio::select! {
        _ = kill_rx => {
            info!("Killing process");
            if let Err(err) = process.kill().await {
                error!("Failed to kill process: {err:?}");
            }
            (true, process.wait().await)
        }
        result = process.wait() => {
            (false, result)
        }
    };
    if killed {
        info!("Killed");
    }
    (killed, result.map_err(Into::into))
}

async fn redirect(mut input: impl AsyncRead + Unpin, to: String) {
    async fn consume(input: &mut (impl AsyncRead + Unpin)) -> Result<()> {
        let mut buffer = [0u8; 2048];
        loop {
            let n = input.read(&mut buffer).await?;
            if n == 0 {
                return Ok(());
            }
        }
    }
    if let Err(e) = try_redirect(&mut input, to).await {
        error!("Failed to redirect process output: {e}");
    }
    if let Err(e) = consume(&mut input).await {
        error!("Failed to consume process output: {e}");
    }
}

async fn try_redirect(input: &mut (impl AsyncRead + Unpin), to: String) -> Result<()> {
    let dst_path = Path::new(&to);
    let dst_path_buf = dst_path.to_path_buf();
    let (reopen_tx, mut reopen_rx) = mpsc::channel(1);

    // Set up file system watcher for logrotate detection
    let mut watcher =
        notify::recommended_watcher(move |res: Result<notify::Event, notify::Error>| {
            if let Ok(event) = res {
                // Check if the event affects our specific file
                if (event.kind.is_remove() || event.kind.is_modify())
                    && event.paths.iter().any(|p| p == &dst_path_buf)
                {
                    let _ = reopen_tx.blocking_send(());
                }
            }
        })?;
    // Watch the log file's parent directory
    watcher.watch(
        dst_path.parent().unwrap_or(Path::new(".")),
        RecursiveMode::NonRecursive,
    )?;

    let mut buffer = [0u8; 8192];
    loop {
        // Open or reopen the log file in append mode
        let mut file = fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(dst_path)?;

        loop {
            tokio::select! {
                // Handle reopening signal
                _ = reopen_rx.recv() => {
                    // Sync file to ensure all data is written
                    if let Err(e) = file.sync_all() {
                        error!("Failed to sync log file: {e}");
                        break;
                    }
                    break; // Break inner loop to reopen file
                }
                // Read and write data
                result = input.read(&mut buffer) => {
                    match result {
                        Ok(0) => return Ok(()), // EOF
                        Ok(n) => {
                            if let Err(e) = file.write_all(&buffer[..n]) {
                                error!("Failed to write to log file: {e}");
                                break;
                            }
                        }
                        Err(e) => {
                            error!("Failed to read from process: {e}");
                            return Err(e.into());
                        }
                    }
                }
            }
        }
    }
}
