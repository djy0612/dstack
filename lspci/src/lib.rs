use std::process::Command;

use anyhow::{Context, Result};

/// Represents a PCI device with the specified fields.
#[derive(Debug)]
pub struct Device {
    pub slot: String,
    pub class: String,
    pub class_id: String,
    pub description: String,
    pub vendor_id: String,
    pub product_id: String,
    pub control: Control,
    pub numa_node: Option<u32>,
}

/// Represents the control settings of a PCI device.
#[derive(Debug)]
pub struct Control {
    pub io: bool,
    pub mem: bool,
    pub bus_master: bool,
}

impl Device {
    pub fn full_product_id(&self) -> String {
        format!("{}:{}", self.vendor_id, self.product_id)
    }

    pub fn in_use(&self) -> bool {
        self.control.bus_master
    }
}

/// Runs `lspci` and parses the output into a vector of `Device` structs.
pub fn lspci_filtered(filter: impl Fn(&Device) -> bool) -> Result<Vec<Device>> {
    let output = Command::new("lspci")
        .args(["-nn", "-vv"])
        .output()
        .context("Failed to run lspci")?;
    Ok(parse_lspci(
        &String::from_utf8(output.stdout).context("Failed to parse lspci output")?,
        filter,
    ))
}

/// Parses the `lspci` output into a vector of `Device` structs.
///
/// # Arguments
/// * `output` - A string slice containing the `lspci` output.
///
/// # Returns
/// A vector of `Device` structs, each representing a parsed PCI device.
pub fn parse_lspci(output: &str, filter: impl Fn(&Device) -> bool) -> Vec<Device> {
    let lines: Vec<&str> = output.lines().collect();
    let mut devices = Vec::new();
    let mut current_device = Vec::new();

    // Group lines into device sections
    for line in lines {
        let line = line.trim_end();
        if !line.is_empty() {
            if !line.starts_with("\t") {
                if !current_device.is_empty() {
                    let device = parse_device(&current_device);
                    if filter(&device) {
                        devices.push(device);
                    }
                    current_device = Vec::new();
                }
                current_device.push(line);
            } else {
                current_device.push(line.trim_start()); // Remove leading tab
            }
        }
    }
    if !current_device.is_empty() {
        let device = parse_device(&current_device);
        if filter(&device) {
            devices.push(device);
        }
    }
    devices
}

/// Parses a single device section into a `Device` struct.
///
/// # Arguments
/// * `device_lines` - A slice of strings representing the lines for one device.
///
/// # Returns
/// A `Device` struct with the parsed fields.
fn parse_device(device_lines: &[&str]) -> Device {
    let device_line = device_lines[0];
    let (slot, class, class_id, description, vendor_id, product_id) =
        parse_device_line(device_line);

    let mut control = Control {
        io: false,
        mem: false,
        bus_master: false,
    };
    let mut numa_node = None;

    // Parse detail lines
    for line in device_lines.iter().skip(1) {
        if line.starts_with("Control: ") {
            control = parse_control_line(line);
        } else if line.starts_with("NUMA node: ") {
            numa_node = parse_numa_node_line(line);
        }
    }

    Device {
        slot,
        class,
        class_id,
        description,
        vendor_id,
        product_id,
        control,
        numa_node,
    }
}

/// Parses the main device line to extract slot, class, class_id, description, vendor_id, and product_id.
///
/// # Arguments
/// * `line` - The main line of a device entry (e.g., "ff:1e.5 System peripheral [0880]: ...").
///
/// # Returns
/// A tuple of strings containing the parsed fields.
fn parse_device_line(line: &str) -> (String, String, String, String, String, String) {
    let words: Vec<&str> = line.split_whitespace().collect();
    if let Some(i) = words.iter().position(|w| w.ends_with("]:")) {
        let slot = words[0].to_string();
        let class = words[1..i].join(" ");
        let class_id = &words[i][1..words[i].len() - 2]; // Remove [ and ]:

        // Look for the vendor ID and product ID pattern [vendor_id:product_id]
        if let Some(j) = words[i + 1..]
            .iter()
            .position(|w| w.starts_with("[") && w.contains(":") && w.ends_with("]"))
        {
            // Extract vendor name without including the "Device" keyword
            let vendor_end = i + 1 + j;
            let mut vendor_words = Vec::new();

            #[allow(clippy::needless_range_loop)]
            for k in i + 1..vendor_end {
                // Skip the word "Device" if it's standalone
                if words[k] != "Device" {
                    vendor_words.push(words[k]);
                }
            }

            let description = vendor_words.join(" ");
            let id_part = &words[vendor_end][1..words[vendor_end].len() - 1]; // Remove [ and ]

            if let Some((vendor_id, product_id)) = id_part.split_once(':') {
                return (
                    slot,
                    class,
                    class_id.to_string(),
                    description,
                    vendor_id.to_string(),
                    product_id.to_string(),
                );
            }
        }
    }
    // Return defaults if parsing fails
    (
        String::new(),
        String::new(),
        String::new(),
        String::new(),
        String::new(),
        String::new(),
    )
}

/// Parses the control line to extract io, mem, and bus_master settings.
///
/// # Arguments
/// * `line` - The control line (e.g., "Control: I/O- Mem- BusMaster- ...").
///
/// # Returns
/// A `Control` struct with the parsed settings.
fn parse_control_line(line: &str) -> Control {
    let mut control = Control {
        io: false,
        mem: false,
        bus_master: false,
    };
    if let Some(items) = line.strip_prefix("Control: ") {
        for item in items.split_whitespace() {
            match item {
                "I/O+" => control.io = true,
                "I/O-" => control.io = false,
                "Mem+" => control.mem = true,
                "Mem-" => control.mem = false,
                "BusMaster+" => control.bus_master = true,
                "BusMaster-" => control.bus_master = false,
                _ => {}
            }
        }
    }
    control
}

/// Parses the NUMA node line to extract the node number.
///
/// # Arguments
/// * `line` - The NUMA node line (e.g., "NUMA node: 1").
///
/// # Returns
/// An `Option<u32>` with the parsed NUMA node number, or `None` if not applicable.
fn parse_numa_node_line(line: &str) -> Option<u32> {
    if let Some(numa_str) = line.strip_prefix("NUMA node: ") {
        numa_str.parse::<u32>().ok()
    } else {
        None
    }
}

#[test]
fn test_lspci() {
    let lspci_output = r#"
bc:02.0 PCI bridge [0604]: PMC-Sierra Inc. Device [11f8:4128] (prog-if 00 [Normal decode])
	Subsystem: NVIDIA Corporation Device [10de:1643]
	Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr+ Stepping- SERR+ FastB2B- DisINTx+
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Latency: 0, Cache Line Size: 32 bytes
	Interrupt: pin ? routed to IRQ 92
	NUMA node: 1
	IOMMU group: 82
	Bus: primary=bc, secondary=bf, subordinate=bf, sec-latency=0
	I/O behind bridge: 0000f000-00000fff [disabled] [32-bit]
	Memory behind bridge: de000000-dfffffff [size=32M] [32-bit]
	Prefetchable memory behind bridge: 00000000fff00000-00000000000fffff [disabled] [64-bit]
	Secondary status: 66MHz- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- <SERR- <PERR-
	BridgeCtl: Parity+ SERR+ NoISA- VGA- VGA16- MAbort- >Reset- FastB2B-
		PriDiscTmr- SecDiscTmr- DiscTmrStat- DiscTmrSERREn-
	Capabilities: <access denied>
	Kernel driver in use: pcieport

bc:03.0 PCI bridge [0604]: PMC-Sierra Inc. Device [11f8:4128] (prog-if 00 [Normal decode])
	Subsystem: NVIDIA Corporation Device [10de:1643]
	Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr+ Stepping- SERR+ FastB2B- DisINTx+
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Latency: 0, Cache Line Size: 32 bytes
	Interrupt: pin ? routed to IRQ 93
	NUMA node: 1
	IOMMU group: 82
	Bus: primary=bc, secondary=c0, subordinate=c0, sec-latency=0
	I/O behind bridge: 0000f000-00000fff [disabled] [32-bit]
	Memory behind bridge: dc000000-ddffffff [size=32M] [32-bit]
	Prefetchable memory behind bridge: 00000000fff00000-00000000000fffff [disabled] [64-bit]
	Secondary status: 66MHz- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- <SERR- <PERR-
	BridgeCtl: Parity+ SERR+ NoISA- VGA- VGA16- MAbort- >Reset- FastB2B-
		PriDiscTmr- SecDiscTmr- DiscTmrStat- DiscTmrSERREn-
	Capabilities: <access denied>
	Kernel driver in use: pcieport

bd:00.0 Bridge [0680]: NVIDIA Corporation GH100 [H100 NVSwitch] [10de:22a3] (rev a1)
	Subsystem: NVIDIA Corporation GH100 [H100 NVSwitch] [10de:1796]
	Physical Slot: 1-1
	Control: I/O- Mem- BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr+ Stepping- SERR+ FastB2B- DisINTx-
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Interrupt: pin A routed to IRQ 10
	NUMA node: 1
	IOMMU group: 82
	Region 0: Memory at e2000000 (64-bit, non-prefetchable) [disabled] [size=32M]
	Capabilities: <access denied>

be:00.0 Bridge [0680]: NVIDIA Corporation GH100 [H100 NVSwitch] [10de:22a3] (rev a1)
	Subsystem: NVIDIA Corporation GH100 [H100 NVSwitch] [10de:1796]
	Physical Slot: 2-1
	Control: I/O- Mem- BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr+ Stepping- SERR+ FastB2B- DisINTx-
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Interrupt: pin A routed to IRQ 5
	NUMA node: 1
	IOMMU group: 82
	Region 0: Memory at e0000000 (64-bit, non-prefetchable) [disabled] [size=32M]
	Capabilities: <access denied>

bf:00.0 Bridge [0680]: NVIDIA Corporation GH100 [H100 NVSwitch] [10de:22a3] (rev a1)
	Subsystem: NVIDIA Corporation GH100 [H100 NVSwitch] [10de:1796]
	Physical Slot: 3-1
	Control: I/O- Mem- BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr+ Stepping- SERR+ FastB2B- DisINTx-
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Interrupt: pin A routed to IRQ 11
	NUMA node: 1
	IOMMU group: 82
	Region 0: Memory at de000000 (64-bit, non-prefetchable) [disabled] [size=32M]
	Capabilities: <access denied>
	IOMMU group: 468
"#;

    let devices = parse_lspci(lspci_output, |_| true);
    insta::assert_debug_snapshot!(devices);
    assert_eq!(devices[0].full_product_id(), "11f8:4128");
}
