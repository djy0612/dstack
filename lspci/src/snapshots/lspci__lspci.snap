---
source: lspci/src/lib.rs
assertion_line: 303
expression: devices
---
[
    Device {
        slot: "bc:02.0",
        class: "PCI bridge",
        class_id: "0604",
        description: "PMC-Sierra Inc.",
        vendor_id: "11f8",
        product_id: "4128",
        control: Control {
            io: true,
            mem: true,
            bus_master: true,
        },
        numa_node: Some(
            1,
        ),
    },
    Device {
        slot: "bc:03.0",
        class: "PCI bridge",
        class_id: "0604",
        description: "PMC-Sierra Inc.",
        vendor_id: "11f8",
        product_id: "4128",
        control: Control {
            io: true,
            mem: true,
            bus_master: true,
        },
        numa_node: Some(
            1,
        ),
    },
    Device {
        slot: "bd:00.0",
        class: "Bridge",
        class_id: "0680",
        description: "NVIDIA Corporation GH100 [H100 NVSwitch]",
        vendor_id: "10de",
        product_id: "22a3",
        control: Control {
            io: false,
            mem: false,
            bus_master: false,
        },
        numa_node: Some(
            1,
        ),
    },
    Device {
        slot: "be:00.0",
        class: "Bridge",
        class_id: "0680",
        description: "NVIDIA Corporation GH100 [H100 NVSwitch]",
        vendor_id: "10de",
        product_id: "22a3",
        control: Control {
            io: false,
            mem: false,
            bus_master: false,
        },
        numa_node: Some(
            1,
        ),
    },
    Device {
        slot: "bf:00.0",
        class: "Bridge",
        class_id: "0680",
        description: "NVIDIA Corporation GH100 [H100 NVSwitch]",
        vendor_id: "10de",
        product_id: "22a3",
        control: Control {
            io: false,
            mem: false,
            bus_master: false,
        },
        numa_node: Some(
            1,
        ),
    },
]
